File: src/fable/Fable.Core/ts/RegExp.ts
type MatchEvaluator = (match: any) => string;

export function create(pattern: string, options: number) {
  let flags = "g";
  flags += options & 1 ? "i" : "";
  flags += options & 2 ? "m" : "";
  return new RegExp(pattern, flags);
}
// From http://stackoverflow.com/questions/3446170/escape-string-for-use-in-javascript-regex

export function escape(str: string) {
  return str.replace(/[\-\[\/\{\}\(\)\*\+\?\.\\\^\$\|]/g, "\\$&");
}

export function unescape(str: string) {
  return str.replace(/\\([\-\[\/\{\}\(\)\*\+\?\.\\\^\$\|])/g, "$1");
}

export function isMatch(str: string | RegExp, pattern: string, options: number = 0) {
  var reg: RegExp = str instanceof RegExp
    ? (reg = <RegExp>str, str = pattern, reg.lastIndex = options, reg)
    : reg = create(pattern, options);
  return reg.test(<string>str);
}

export function match(str: string | RegExp, pattern: string, options: number = 0) {
  var reg: RegExp = str instanceof RegExp
    ? (reg = <RegExp>str, str = pattern, reg.lastIndex = options, reg)
    : reg = create(pattern, options);
  return reg.exec(<string>str);
}

export function matches(str: string | RegExp, pattern: string, options: number = 0) {
  var reg: RegExp = str instanceof RegExp
    ? (reg = <RegExp>str, str = pattern, reg.lastIndex = options, reg)
    : reg = create(pattern, options);
  if (!reg.global)
    throw "Non-global RegExp"; // Prevent infinite loop
  let m: RegExpExecArray;
  const matches: RegExpExecArray[] = [];
  while ((m = reg.exec(<string>str)) !== null)
    matches.push(m);
  return matches;
}

export function options(reg: RegExp) {
  let options = 256; // ECMAScript
  options |= reg.ignoreCase ? 1 : 0;
  options |= reg.multiline ? 2 : 0;
  return options;
}

export function replace(reg: string | RegExp, input: string, replacement: string | MatchEvaluator, limit?: number, offset: number = 0) {
  function replacer() {
    let res = arguments[0];
    if (limit !== 0) {
      limit--;
      const match: any = [];
      const len = arguments.length;
      for (let i = 0; i < len - 2; i++)
        match.push(arguments[i]);
      match.index = arguments[len - 2];
      match.input = arguments[len - 1];
      res = (replacement as MatchEvaluator)(match);
    }
    return res;
  }
  if (typeof reg == "string") {
    const tmp = <string>reg;
    reg = create(input, limit);
    input = tmp;
    limit = undefined;
  }
  if (typeof replacement == "function") {
    limit = limit == null ? -1 : limit;
    return input.substring(0, offset) + input.substring(offset).replace(<RegExp>reg, replacer);
  } else {
    if (limit != null) {
      let m: RegExpExecArray;
      const sub1 = input.substring(offset);
      const _matches = matches(reg, sub1);
      const sub2 = matches.length > limit ? (m = _matches[limit - 1], sub1.substring(0, m.index + m[0].length)) : sub1;
      return input.substring(0, offset) + sub2.replace(<RegExp>reg, <string>replacement) + input.substring(offset + sub2.length);
    } else {
      return input.replace(<RegExp>reg, <string>replacement);
    }
  }
}

export function split(reg: string | RegExp, input: string, limit?: number, offset: number = 0) {
  if (typeof reg == "string") {
    const tmp = <string>reg;
    reg = create(input, limit);
    input = tmp;
    limit = undefined;
  }
  input = input.substring(offset);
  return input.split(<RegExp>reg, limit);
}

File: src/fable-library-pyx/fable_library/reg_exp.py
import re

from typing import (
    Callable,
    Iterator,
    List,
    Match,
    Optional,
    Pattern,
    Union,
    Dict,
)


MatchEvaluator = Callable[[Match[str]], str]


class GroupCollection:
    def __init__(self, groups: List[str], named_groups: Dict[str, str]) -> None:
        self.named_groups = named_groups
        self.groups = groups

    def values(self) -> List[str]:
        return list(self.groups)

    def __len__(self) -> int:
        return len(self.groups)

    def __getitem__(self, key: Union[int, str]) -> Optional[str]:
        if isinstance(key, int):
            return self.groups[key]
        else:
            return self.named_groups.get(key)

    def __iter__(self) -> Iterator[str]:
        return iter(self.groups)


def _options_to_flags(options: int) -> int:
    """Convert .NET options to Python flags."""
    flags = 256
    if options & 1:
        flags |= re.IGNORECASE
    if options & 2:
        flags |= re.MULTILINE
    if options & 16:
        flags |= re.S
    return flags


def create(pattern: str, options: int = 0) -> Pattern[str]:
    # raw_pattern = pattern.encode("unicode_escape").decode("utf-8")
    flags = _options_to_flags(options)
    pattern = pattern.replace("?<", "?P<")
    return re.compile(pattern, flags=flags)


def escape(string: str) -> str:
    return re.escape(string)


def unescape(string: str) -> str:
    return re.sub(r"\\(.)", r"\1", string)


def match(
    reg: Union[Pattern[str], str], input: str, start_at: int = 0
) -> Optional[Match[str]]:
    if isinstance(reg, str):
        flags = _options_to_flags(start_at)
        return re.search(input, reg, flags)
    return reg.search(input, pos=start_at)


def matches(reg: Pattern[str], input: str, start_at: int = 0) -> List[Match[str]]:
    if isinstance(reg, str):
        flags = _options_to_flags(start_at)
        input = input.replace("?<", "?P<")
        return list(re.finditer(input, reg, flags=flags))

    return list(reg.finditer(input, pos=start_at))


def is_match(reg: Union[Pattern[str], str], input: str, start_at: int = 0) -> bool:
    if isinstance(reg, str):
        # Note: input is the pattern here
        flags = _options_to_flags(start_at)
        return re.search(input, reg, flags=flags) is not None

    return reg.search(input, pos=start_at) is not None


def groups(m: Match[str]) -> GroupCollection:
    named_groups: Dict[str, str] = m.groupdict()

    # .NET adds the whole capture as group 0
    groups_ = [m.group(0), *m.groups()]

    return GroupCollection(named_groups=named_groups, groups=groups_)


def options(reg: Pattern[str]) -> int:
    options = 256  # ECMAScript
    if reg.flags & re.IGNORECASE:
        options |= 1
    if reg.flags & re.MULTILINE:
        options |= 2
    if reg.flags & re.S:
        options |= 16
    return options


def replace(
    reg: Union[str, Pattern[str]],
    input: str,
    replacement: Union[str, MatchEvaluator],
    limit: Optional[int] = None,
    offset: int = 0,
) -> str:
    if isinstance(replacement, str):
        # Rewrite placeholders from .NET e.g $1 to \g<1> Python syntax
        replacement = re.sub(pattern=r"\$(\d+)", repl=r"\\g<\g<1>>", string=replacement)

    if isinstance(reg, str):
        return re.sub(input, replacement, reg, count=limit or 0)
    else:
        return input[:offset] + reg.sub(replacement, input[offset:], count=limit or 0)


def split(
    reg: Union[str, Pattern[str]],
    input: str,
    limit: Optional[int] = None,
    offset: int = 0,
) -> List[str]:
    if isinstance(reg, str):
        return re.split(input, reg, maxsplit=limit or 0)

    input = input[offset:]
    return reg.split(input, maxsplit=limit or 0)[:limit]


def get_item(groups: GroupCollection, index: Union[str, int]) -> Optional[str]:
    return groups[index]


__all__ = [
    "escape",
    "is_match",
    "match",
    "matches",
    "create",
    "options",
    "replace",
    "split",
    "unescape",
    "groups",
    "get_item",
]
